File: src/main/java/dev/bluesheep/xeiexporter/recipe/ShapelessRecipeExporter.java
package dev.bluesheep.xeiexporter.recipe;

import com.google.gson.JsonArray;
import com.google.gson.JsonObject;
import com.mojang.serialization.JsonOps;
import net.minecraft.nbt.CompoundTag;
import net.minecraft.world.item.ItemStack;
import net.minecraft.world.item.crafting.ShapelessRecipe;
import net.minecraft.world.level.Level;
import net.minecraftforge.registries.ForgeRegistries;

import java.util.Objects;

public class ShapelessRecipeExporter {
    ShapelessRecipeExporter() {}

    public JsonObject export(ShapelessRecipe recipe, Level level) {
        JsonObject json = new JsonObject();
        json.addProperty("type", "minecraft:crafting_shapeless");
        json.addProperty("category", recipe.category().getSerializedName());
        if (!recipe.getGroup().isEmpty()) {
            json.addProperty("group", recipe.getGroup());
        }

        JsonArray ingredients = new JsonArray();
        recipe.getIngredients().forEach(it -> ingredients.add(it.toJson()));
        json.add("ingredients", ingredients);

        ItemStack itemStack = recipe.getResultItem(level.registryAccess());
        JsonObject result = new JsonObject();
        result.addProperty("item", Objects.requireNonNull(ForgeRegistries.ITEMS.getKey(itemStack.getItem())).toString());
        result.addProperty("count", itemStack.getCount());
        CompoundTag.CODEC.encodeStart(JsonOps.INSTANCE, itemStack.getTag())
                .result()
                .ifPresent(tag -> result.add("tag", tag));
        json.add("result", result);

        return json;
    }
}

File: src/main/java/dev/bluesheep/xeiexporter/exporter/recipe/ShapelessRecipeExporter.kt
package dev.bluesheep.xeiexporter.exporter.recipe

import dev.bluesheep.xeiexporter.api.recipe.IRecipeExporter
import dev.bluesheep.xeiexporter.api.recipe.RecipeData
import dev.bluesheep.xeiexporter.api.recipe.ingredient.ItemRecipeIngredient
import dev.bluesheep.xeiexporter.api.recipe.result.ItemRecipeResult
import dev.bluesheep.xeiexporter.exporter.ExportUtil.rlVanilla
import net.minecraft.resources.ResourceLocation
import net.minecraft.world.item.crafting.ShapelessRecipe
import net.minecraft.world.level.Level

class ShapelessRecipeExporter() : IRecipeExporter<ShapelessRecipe> {
    override val recipeTypeId: ResourceLocation = rlVanilla("crafting_shapeless")

    override fun export(recipe: ShapelessRecipe, level: Level): RecipeData {
        val ingredients = recipe.ingredients.map { ItemRecipeIngredient(it) }
        return RecipeData(
            recipe.id,
            recipeTypeId,
            ingredients,
            listOf(ItemRecipeResult(recipe.getResultItem(level.registryAccess())))
        )
    }
}
